refactor: 채팅기능 개선 + 채팅방 CRUD 구현

lss6181 opened the pull request:
## Issue #1 , refactor Pr #21 
## 변경사항
- `MessageDto` ➡️ `MessageRequestDto` / `MessageResponseDto` 두개로 구분시켜 사용
- `ErrorCode.CHAT_NOT_MY_GYM`, `ErrorCode.NEED_ENTER` 추가 + `C004~C009` 추가
- `GroupChatMember` Entity ➡️ `UserChat`으로 변경
- `User` Entity :  `groupChatMemberList` ➡️ `userChatList` 이름변경
```java
    @OneToMany(mappedBy = "user", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<UserChat> userChatList = new ArrayList<>();
```
- `Chat` Entity :  `User` 필드 `@OneToOne` 매핑으로 채팅방 생성한 주인 값 담을 수 있도록 추가
```java
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User ownerUser;
```
- `ChatController`, `ChatService` : 채팅방 CRUD + 채팅방에 참여/나가기 API 구현
- 채팅방 참여하는 API요청("/chats/{chatId}/joiners") 후 채팅방에 입장(소켓 ENTER요청) 가능하도록 설정
```java
    /*
     * Json형태의 입력값과 해당 클라이언트 세션 받아와서 로직수행을 시작 하는 곳.
     * MessageRequestDto필드대로 입력해준다.
     */
    @Override
    protected void handleTextMessage(WebSocketSession session, TextMessage message)
        throws Exception {
        .
        .
        .
        // 채팅방 참여완료 했는지 (UserChat에 있는지) check
        chatService.validateJoinedChat(sendUser, chat);
        .
        .
    }

    public void validateJoinedChat(User user, Chat chat) {
        if (!userChatRepository.existsByChatAndUser(chat, user)) {
            throw new CustomException(ErrorCode.USERCHAT_NOT_FOUND);
        }
    }
```
- `WebSocketHandler`, `ChatService` : 비효율적인 클래스주입이나 중복되는 코드 메서드화, 가독성 향상을 위한 코드 개선
## 세션유지가 안됐던 이전 코드 개선
- 은비님 브랜치 참조하여 `WebSocketHandler`에 있던 `sessions` ➡️ `SessionManager`클래스 생성하여 클래스 주입형태로 바꿈 (추 후 session을 `Map`으로 관리하며 `roomId`를 Key값으로 하신것도 참고해서 수정 해볼 예정)
  - ↪️`SessionManager` : `Map<Long, Set<WebSocketSession>>`인 sessions로 변경 하여 session운영 적용 완료
```java
private Map<Long, Set<WebSocketSession>> sessions = new HashMap<>();
```
- `ENTER` 입장 : `SessionManager.sessions`에 해당 session 추가해주는 `SessionManager.addSession(Long chatId, WebSocketSession session)` 구현.  입장한 채팅방을 `chatId`로 구분하여 키값으로 활용한다.
```java
    public void addSession(Long chatId, WebSocketSession session) {
        this.sessions
            .computeIfAbsent(chatId,k -> new HashSet<>())
            .add(session);
    }

computeIfAbsent() 메서드는 다음과 같이 동작합니다.

Key가 존재할 경우: 아무런 작업을 하지 않고 기존에 존재하는 Key의 Value를 리턴
Key가 존재하지 않는 경우: 람다식을 적용한 값을 해당 key에 저장한 후 newValue를 반환
```

- 채팅방 연결 종료 : 종료한 클라이언트의 `session` ➡️ `sessionManager.deleteSessionAndGetChatId(session)` + 채팅방 나갔다는 알림 메세지 전송
```java
    public Long deleteSessionAndGetChatId(WebSocketSession session) {
        Long chatId = null;
        for (Map.Entry<Long, Set<WebSocketSession>> entry : sessions.entrySet()) {
            Long id = entry.getKey();
            Set<WebSocketSession> set = entry.getValue();

            if (set.contains(session)) {
                set.remove(session);
                chatId = id;
            }
        }
        return chatId;
    }
```
- `ENTER` 요청 과정 없이(=채팅방 입장 없이) `TALK` 요청으로 메세지 send 시, 입장먼저 필요하다는 예외 throw
```java
    private void checkEnterIsPresent(WebSocketSession session) {
        if (session.getAttributes().isEmpty()) {    // ENTER(입장) 없이 TALK 요청 할 경우. ENTER를 했다면 getAttributes()했을 때 userId=1 이런식으로 있어야됨. 
            throw new CustomException(ErrorCode.NEED_ENTER);
        }
    }
```
- `ENTER` 요청 완료 후 중복 `ENTER` 요청 시, 이미 입장했다는 메세지 발송. 해당 메세지는 본인 에게만 발송됨.
```java
    private void alreadyExistEnter(WebSocketSession session, MessageRequestDto chatMessage){

        // 이미 입장했다는 알림으로 바꾸기.
        chatMessage.changeAlreadyEntered();
        // 본인에게만 sendMessage
        chatService.sendMessage(session, chatMessage);
    }
```
- `TALK` 메세지 발송 : 메세지 발송 한 `chatId` ➡️ `SessionManager.sessions`에서 키값이 `chatId` 인 value(`Set<WebSocketSession>`)를 가져와 해당set에 있는  session들에게 메세지 보여줌. 
  - 즉, 메세지 보낸 채팅방의 참여중인 인원들에게만 메세지가 보내지도록 구현.
```java
public Set<WebSocketSession> getSessionsByChatId(Long chatId) {
        return sessions.getOrDefault(chatId, Collections.emptySet());
    }
```

lss6181 committed:
feat: 채팅방 생성 기능, 웹소켓 미완성 일부 설정(dependency 추가, config 설정)

lss6181 committed:
Merge branch 'develop' of https://github.com/Body-Buddy/BodyBuddy-BE into feature/1-chat

lss6181 committed:
feat: Users -> User로 바뀐 부분들 수정

lss6181 committed:
feat: 채팅기능
- "ws://localhost:8080/ws/chat" 으로 웹소켓 요청 (with 크롬 확장프로그램 WebSocketTestClient)
- Json타입으로 요청함.
- MessageType ENTER(채팅방 입장), TALK(현재 참여중) 구분
    - ENTER일 경우 각 채팅방인원(session)들에게 입장했다는 메세지 발송됨.
    - TALK일 경우 메세지 띄워주고 MessageRepository에 해당 메세지 저장
- 채팅방 입장 시 해당 채팅방의 이전 대화내용 불러와 보여줌.
- User.nickname -> (nullable = false, unique = true) 추가
- UserRepository.findByNickname 메서드 추가
- ErrorCode.CHAT_NOT_FOUND, USER_NOT_FOUND 추가

lss6181 committed:
feat: 채팅
- UserGymRepository.existsByUserAndGym() 메서드 활용하여 등록한 짐에서만 채팅방 생성 가능하도록 로직생성
- MessageDto.chatId String으로 변경

lss6181 committed:
Merge branch 'develop' of https://github.com/Body-Buddy/BodyBuddy-BE into feature/1-chat

# Conflicts:
#	src/main/java/com/bb3/bodybuddybe/common/exception/ErrorCode.java
#	src/main/java/com/bb3/bodybuddybe/user/repository/UserRepository.java

lss6181 committed:
style: ErrorCode.CHAT_NOT_FOUND, ErrorCode.USER_NOT_FOUND 코드명 수정

lss6181 committed:
refactor: 채팅기능
- WebSocketHandler 전체적인 리팩터링 진행 (세션유지가 안됐던 이전 코드 개선)
    - HashSet인 sessions 필드 정의 추가. 기존 ChatResponseDto에서 선언 했던 것 -> WebSocketHandler에 선언하여 session 관리
    - ENTER 입장 : sessions에 해당 session add, sessions에 있는 사용자(session)들에게 입장알림 메세지 발송
    - 각 클라이언트 웹 소켓 연결 종료 시 해당 클라이언트 session -> sessions.remove() + 채팅방 나갔다는 알림 메세지 전송
    - ENTER 요청 과정 없이(=채팅방 입장 없이) TALK 요청으로 메세지 send 시, 입장먼저 필요하다는 예외 throw
    - ENTER 요청 후 같은 클라이언트에서 중복 ENTER 요청 시, 이미 입장했다는 메세지 발송. 해당 메세지는 본인 에게만 발송되도록 처리.
    - TALK 메세지 발송 : sessions에 추가된 모든 클라이언트(session)들 에게 메세지 발송
- ErrorCode.CHAT_NOT_MY_GYM, ErrorCode.NEED_ENTER 추가
- MessageDto -> MessageRequestDto / MessageResponseDto 로 구분 하여 별도 활용

lss6181 committed:
Merge branch 'develop' of https://github.com/Body-Buddy/BodyBuddy-BE into feature/1-chat

# Conflicts:
#	src/main/java/com/bb3/bodybuddybe/chat/advice/WebSocketHandler.java
#	src/main/java/com/bb3/bodybuddybe/chat/dto/ChatResponseDto.java
#	src/main/java/com/bb3/bodybuddybe/chat/dto/MessageRequestDto.java
#	src/main/java/com/bb3/bodybuddybe/chat/dto/MessageResponseDto.java
#	src/main/java/com/bb3/bodybuddybe/chat/entity/GroupChatMember.java
#	src/main/java/com/bb3/bodybuddybe/chat/entity/Message.java
#	src/main/java/com/bb3/bodybuddybe/chat/service/ChatService.java
#	src/main/java/com/bb3/bodybuddybe/common/exception/ErrorCode.java
#	src/main/java/com/bb3/bodybuddybe/user/entity/User.java
#	src/main/java/com/bb3/bodybuddybe/user/repository/UserRepository.java

eunb1 commented:
수고 많으셨어요 성수님 🎉
다만 유저 입퇴장 시 유저와 채팅방의 중간 엔티티를 저장 또는 삭제하는 로직이 추가적으로 필요할 것 같아요 !
기존의 `GroupChatMember` 가 그 역할을 하는 것 같은데, 일대일 채팅 기능도 구현하는 만큼 `UserChat`으로 클래스명 변경하는건 어떨까요~~?

lss6181 committed:
refactor: GroupChatMember -> UserChat으로 변경, ChatService.saveUserChat 생성(임시, 수정예정)

lss6181 committed:
refactor: WebSocketHandler에 있던 sessions -> SessionManager클래스 생성하여 클래스 주입으로 sessions관리

lss6181 committed:
Merge branch 'develop' of https://github.com/Body-Buddy/BodyBuddy-BE into feature/1-chat

lss6181 committed:
feat: 채팅방 CRUD 구현 및 개선
- Chat : User와 @OneToOne 매핑으로 채팅방 생성한 주인 값 담을 수 있도록 추가
- ChatController, ChatService : 채팅방 CRUD + 채팅방에 참여/나가기 API 구현
- ErrorCode : C004~C009 추가
- User : 매핑되있던 groupChatMemberList -> userChatList 이름변경

lss6181 committed:
style: 생성자 어노테이션 제거

lss6181 committed:
style: 미사용 import 제거

lss6181 committed:
refactor: 채팅기능 개선
- 채팅방 참여하는 API요청("/chats/{chatId}/joiners") 후 채팅방에 입장(소켓 ENTER요청) 가능하도록 설정 -> chatService.validateJoinedChat()
- SessionManager : Map<Long, Set<WebSocketSession>>인 sessions로 변경 하여 session운영 적용
- 그밖에 비효율적인 클래스주입이나 중복되는 코드 메서드화, 가독성 향상을 위한 코드 개선(WebSocketHandler, ChatService)
